### PR TITLE
dsv4.1: Top-k kernels and candidate selection

### PR DESCRIPTION
**Pull request by hnyls2002**
## Summary
- Add the DeepSeek-V4.1 sparse-indexer kernels that sit beside Top-k v2: a bf16 small-row top-k, the two-level indexer's block-max keys and sorted block table, and a cluster-occupancy probe.
- Rework Top-k v2's edge handling and cluster dispatch (details below); the existing suite still passes and two new regression cases pin the fixed `-inf` tie behaviour.

## New kernels
- `topk_bf16_small` (`topk_transform_bf16_small`): bf16 top-k for rows of at most 16384 scores, two radix passes over the raw bf16 bytes, fused with the page-table transform; adapted from DeepSelect's init-window select.
- `block_amax` (`amax8_varlen`): level-one keys of the two-level indexer, the max of each 8-score block with the row's newest block forced to `+inf`.
- `candidate_block_table` (`sort_candidate_blocks`): counting sort of a row's selected blocks over a bitmap plus the page transform, in place.
- `jit/utils/occupancy.py` (`get_max_active_clusters`): asks `cudaOccupancyMaxActiveClusters` how many clusters of a given width fit at a given occupancy; sizes the Top-k v2 persistent pool at JIT time.
- `attention/dsv4/candidate_table.py` hosts the two indexer helpers; `topk.py` keeps only the top-k entry points.

## Top-k v2 changes
- Padding uses NaN instead of `-FLT_MAX`, so inactive entries never displace valid `-inf` scores (`test_topk_v2_negative_infinity_ties`, `test_topk_v2_ragged_negative_infinity`); the padding bin is taken from `extract_coarse_bin(padding_value())`, so the histogram correction does not depend on the platform's NaN encoding.
- `coarse_bin_lower_bound` steps the midpoint up on odd-significand bins (fp32 -> fp16 rounds to nearest even) and returns `+/-inf` at the key-space edges.
- Cluster path: `TopKCluster<N>` for N = 8 and 16, inputs interleaved across ranks, rank 0 owns the histogram and threshold behind release/acquire cluster barriers; the persistent-pool size and the cluster-16 small-batch cap come from the occupancy probe, with per-arch defaults as fallback; `topk_plan_cluster` carries per-arch candidate tables.
- `SUPPORT_CLUSTER` limits the cluster path to sm90/sm100; sm120 and ROCm take the streaming path.
- `plan_topk_v2(static_threshold)`: a negative value selects the heuristic (was `0`); no in-tree caller passes the argument.
- The #32830 mitigation (`__builtin_assume` in the small-batch cluster epilogue) is dropped: the restructured epilogue compiles with nvcc 13.1 through 13.4 for sm_90a, while the previous code still segfaults cicc without it.
- `TieHandleSmem` no longer over-aligns its counters; `topk_v2_plan_is_written` replaces the test kit's use of the private JIT module.

## Verification
- `test_topk_v2.py`: 288 cases pass on SM90 and SM100.
- The new kernels were checked against torch references (block max with `+inf` newest block, sorted block table with page mapping, bf16 top-k); they have no registered tests of their own and are covered by the model's end-to-end tests.
- `bench_topk.py` seeds each configuration and adds `page_size` to the paged sweep; not re-run here.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #35064790417](https://github.com/sgl-project/sglang/actions/runs/35064790417)
Latest PR Test (Extra): :x: [Run #35064790112](https://github.com/sgl-project/sglang/actions/runs/35064790112)
Latest PR Test (AMD ROCm 10): :no_entry_sign: [Run #35064790372](https://github.com/sgl-project/sglang/actions/runs/35064790372)

### CODE REVIEW

File: python/sglang/kernels/jit/csrc/deepseek_v4/topk_v2.cuh [hunk 1]
```diff
@@ -97,18 +96,16 @@ struct TopKPagedParams {
   SGL_DEVICE int32_t* get_output_ptr(uint32_t batch_id) const {
     return page_indices + batch_id * static_cast<int64_t>(topk);
   }
-  SGL_DEVICE int32_t* get_raw_output_ptr(uint32_t batch_id) const {
-    return raw_indices == nullptr ? nullptr : raw_indices + batch_id * static_cast<int64_t>(topk);
+  SGL_DEVICE PageTransform get_transform(uint32_t batch_id) const {
+    return {page_table + batch_id * page_table_stride, page_bits, raw_indices + batch_id * static_cast<int64_t>(topk)};
```
Comment (BBuf) [inline python/sglang/kernels/jit/csrc/deepseek_v4/topk_v2.cuh]: Both pointers are optional depending on the output mode, but `get_transform()` unconditionally performs pointer arithmetic on them. For example, `page_table` is null in the indices-only mode and `raw_indices` is null in the page-table mode. `nullptr + offset` is undefined behavior even if the corresponding field is later eliminated by an `if constexpr`.

Could we preserve the null value here?

```cpp
page_table ? page_table + batch_id * page_table_stride : nullptr
raw_indices ? raw_indices + batch_id * topk : nullptr
```

File: test/registered/kernels/ops/attention/test_topk_v2.py [hunk 2]
```diff
@@ -387,15 +387,56 @@ def test_topk_v2_ragged_window(name: str, rows, k: int, offset_shift: int) -> No
...
-    # that the kernel actually reads (trivial rows read nothing)
-    changed = (scores != before).cpu()
-    for i, (start, length) in enumerate(rows):
-        allowed = torch.zeros(scores.shape[1], dtype=torch.bool)
-        if length > k:
-            allowed[start - start % 4 : start] = True
-        stray = (changed[i] & ~allowed).nonzero().flatten().tolist()
-        assert not stray, f"row {i} ({name}) wrote outside its masked head: {stray[:8]}"
+
+def _assert_topk_values(window, indices, k):
```
Comment (BBuf) [inline test/registered/kernels/ops/attention/test_topk_v2.py:403]: Could we keep the previous stray-write assertion in addition to the new `-inf` cases? It verified that the only legal in-place mutation was the at-most-three aligned prefix columns before a window. This patch changes the `input_start`/padding paths but removes that check, so an out-of-window write can now pass all tests as long as the selected Top-k values remain correct.

**Review by BBuf — COMMENTED** [review]
This changes occupancy probing, cluster-size selection, persistent dispatch, and several plan thresholds, but the PR description says the benchmark was not run. Before merging, could you attach plan-selection and latency results around the relevant boundaries, especially batch 15/16 and sequence lengths around 24K/30K/32K/64K, on the intended Blackwell target and H100 if it is supported?

Please include the baseline/candidate commits and commands. The “14 files are byte-identical” statement also needs updating after `fd6e5f8`.

File: python/sglang/kernels/jit/csrc/deepseek_v4/topk_v2.cuh (hunk 1 above)
Comment (BBuf) [reply python/sglang/kernels/jit/csrc/deepseek_v4/topk_v2.cuh]: Fixed in 263375f167: both optional pointers now preserve nullptr instead of performing pointer arithmetic on it.

File: test/registered/kernels/ops/attention/test_topk_v2.py (hunk 2 above)
Comment (BBuf) [reply test/registered/kernels/ops/attention/test_topk_v2.py:403]: Restored the stray-write assertion while keeping the new -inf coverage. The retained top-k integration test passed 288/288 cases on H200.

**Comment by hnyls2002** [comment]
/rerun-test test/registered/kernels/ops/attention/test_dsa_indexer.py test/registered/kernel/attention/test_dsa_metadata_replay.py test/registered/kernels/ops/attention/test_dsv32_indexer_fusion.py test/registered/unit/layers/test_dsv4_nonpaged_indexer.py
